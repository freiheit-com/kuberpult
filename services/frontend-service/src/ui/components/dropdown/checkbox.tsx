/*This file is part of kuberpult.

Kuberpult is free software: you can redistribute it and/or modify
it under the terms of the Expat(MIT) License as published by
the Free Software Foundation.

Kuberpult is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
MIT License for more details.

You should have received a copy of the MIT License
along with kuberpult. If not, see <https://directory.fsf.org/wiki/License:Expat>.

Copyright 2023 freiheit.com*/

import * as React from 'react';
import { Button } from '../button';

export type CheckboxProps = {
    onClick?: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
    classes?: string;
    id: string;
    enabled: boolean;
    label: string;
};

export const Checkbox: React.FC<CheckboxProps> = (props) => (
    <span onClick={props.onClick} className={'checkbox-wrapper'} id={String(props.id)}>
        <Button
            className={'test-button-checkbox id-' + props.id + ' ' + (props.enabled ? 'enabled' : 'disabled')}
            label={props.enabled ? '☑' : '☐'}
        />
        {props.label}
    </span>
);
