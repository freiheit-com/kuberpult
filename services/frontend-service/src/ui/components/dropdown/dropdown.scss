@use '@material/floating-label/mdc-floating-label';
@use '@material/notched-outline/mdc-notched-outline';
@use '@material/select/mdc-select';
@use '@material/select/icon';

.mdc-select {
  width: 100%;
  fieldset {
    border-color: var(--mdc-theme-primary);
    margin-left: 50px;
    border-radius: 10px;
  }
}
