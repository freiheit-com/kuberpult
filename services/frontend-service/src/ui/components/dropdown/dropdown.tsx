/*This file is part of kuberpult.

Kuberpult is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Kuberpult is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with kuberpult.  If not, see <http://www.gnu.org/licenses/>.

Copyright 2021 freiheit.com*/
import { InputLabel, MenuItem, OutlinedInput, Select } from '@material-ui/core';
import classNames from 'classnames';
import { useRef } from 'react';
import { useTeamNames } from '../../utils/store';

export type DropdownProps = {
    className?: string;
    floatingLabel?: string;
    leadingIcon?: string;
};

export const Dropdown = (props: DropdownProps) => {
    const { className, floatingLabel, leadingIcon } = props;
    const control = useRef<HTMLDivElement>(null);
    const teams = useTeamNames();
    var team = '';
    console.log(teams);
    const allClassName = classNames(
        'mdc-select',
        'mdc-select--outlined',
        {
            'mdc-select--no-label': !floatingLabel,
            'mdc-select--with-leading-icon': leadingIcon,
        },
        className
    );

    return (
        <div className={allClassName} ref={control}>
            <InputLabel className="mdc-floating-label">{floatingLabel}</InputLabel>
            <Select defaultValue={'All'} className="mdc-select">
                <MenuItem value={''}>All</MenuItem>
                {teams.map((el: string) => (
                    <MenuItem value={el}>{el}</MenuItem>
                ))}
            </Select>
        </div>
    );
};
