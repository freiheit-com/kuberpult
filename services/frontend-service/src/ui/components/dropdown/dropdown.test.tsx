/*This file is part of kuberpult.

Kuberpult is free software: you can redistribute it and/or modify
it under the terms of the Expat(MIT) License as published by
the Free Software Foundation.

Kuberpult is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
MIT License for more details.

You should have received a copy of the MIT License
along with kuberpult. If not, see <https://directory.fsf.org/wiki/License:Expat>.

Copyright 2023 freiheit.com*/
import { DropdownSelect } from './dropdown';
import { getByTestId, render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

const getNode = (overrides?: {}): JSX.Element | any => {
    const defaultProps: any = {
        children: null,
    };
    // given
    return (
        <div>
            <MemoryRouter>
                <DropdownSelect {...defaultProps} {...overrides} />
            </MemoryRouter>
        </div>
    );
};

const getWrapper = (
    overrides?: {
        handleChange: (event: any) => void;
        isEmpty: (arr: string[] | undefined) => boolean;
        floatingLabel: string;
        teams: string[];
        selectedTeams: string[];
    },
    entries?: string[]
) => render(getNode(overrides));

describe('Dropdown label', () => {
    interface dataT {
        name: string;
        floatingLabel: string;
        expectedLabel: RegExp;
    }

    const data: dataT[] = [
        {
            name: 'Get label when no teams are selected',
            floatingLabel: 'Teams',
            expectedLabel: /^Teams$/,
        },
        {
            name: 'Get label when no teams are selected',
            floatingLabel: 'Test',
            expectedLabel: /^Test$/,
        },
    ];

    describe.each(data)(`Renders a navigation item with selected`, (testcase) => {
        it(testcase.name, () => {
            // given
            const { floatingLabel } = testcase;
            // when
            const { container } = getWrapper({
                isEmpty: (arr: string[] | undefined) => (arr ? arr.filter((val) => val !== '').length === 0 : true),
                handleChange: (event: any) => {},
                floatingLabel: floatingLabel,
                teams: ['example', 'bar'],
                selectedTeams: [],
            });
            // then
            expect(getByTestId(container, 'teams-dropdown-label')).toHaveTextContent(testcase.expectedLabel);
        });
    });
});

describe('Dropdown shrink label', () => {
    interface dataT {
        name: string;
        selectedTeams: string[];
        expectedShrink: string;
    }

    const data: dataT[] = [
        {
            name: 'Check if label is not shrunk when there are no teams selected',
            selectedTeams: [],
            expectedShrink: 'false',
        },
        {
            name: 'Check if label is shrunk when there are teams selected',
            selectedTeams: ['example', 'bar'],
            expectedShrink: 'true',
        },
    ];

    describe.each(data)(`Renders a navigation item with selected`, (testcase) => {
        it(testcase.name, () => {
            // given
            const { selectedTeams } = testcase;
            // when
            const { container } = getWrapper({
                isEmpty: (arr: string[] | undefined) => (arr ? arr.filter((val) => val !== '').length === 0 : true),
                handleChange: (event: any) => {},
                floatingLabel: 'Teams',
                teams: ['example', 'bar'],
                selectedTeams,
            });
            // then
            expect(getByTestId(container, 'teams-dropdown-label')).toHaveAttribute(
                'data-shrink',
                testcase.expectedShrink
            );
        });
    });
});

describe('Dropdown dropdown text with selected teams', () => {
    interface dataT {
        name: string;
        selectedTeams: string[];
        expectedTeamsText: RegExp;
    }

    const data: dataT[] = [
        {
            name: 'Get value after selecting a team',
            selectedTeams: ['example'],
            expectedTeamsText: /^example$/,
        },
        {
            name: 'Get value after selecting multiple teams',
            selectedTeams: ['example', 'bar'],
            expectedTeamsText: /^example, bar$/,
        },
    ];

    describe.each(data)(`Renders a navigation item with selected`, (testcase) => {
        it(testcase.name, () => {
            // given
            const { selectedTeams } = testcase;
            // when
            const { container } = getWrapper({
                isEmpty: (arr: string[] | undefined) => (arr ? arr.filter((val) => val !== '').length === 0 : true),
                handleChange: (event: any) => {},
                floatingLabel: 'Teams',
                teams: ['example', 'bar'],
                selectedTeams,
            });
            // then
            expect(
                getByTestId(container, 'teams-dropdown-select').getElementsByClassName('MuiSelect-select')[0]
            ).toHaveTextContent(testcase.expectedTeamsText);
            // testcase.expect(container);
        });
    });
});
