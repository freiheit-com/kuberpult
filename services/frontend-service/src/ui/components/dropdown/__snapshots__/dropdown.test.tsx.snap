// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textfield renders correctly using Snapshot 1`] = `
<div
  class="mdc-text-field mdc-text-field--outlined"
>
  <span
    class="mdc-notched-outline mdc-notched-outline--upgraded"
  >
    <span
      class="mdc-notched-outline__leading"
    />
    <span
      class="mdc-notched-outline__notch"
    >
      <span
        class="mdc-floating-label"
        style="transition-duration: 0s;"
      >
        Floating label
      </span>
    </span>
    <span
      class="mdc-notched-outline__trailing"
    />
  </span>
  <input
    aria-label="Floating label"
    class="mdc-text-field__input"
    type="text"
    value=""
  />
</div>
`;
